go/toolchains.rst

```diff
@@ -18,22 +27,568 @@ customize the behavior of the core_ go rules.
+| They can be used to specify target specific flags that c linking actions generated by cgo        |
+| should apply when using this toolchain.                                                          |
++--------------------------------+-----------------------------+-----------------------------------+
+
+The toolchain object
+~~~~~~~~~~~~~~~~~~~~
+
+When you get a go toolchain from a context (see use_) it exposes a number of fields.
+
+The public interface is the actions member.
+This holds a collection of functions for generating the standard actions the toolchain knows
+about, compiling and linking for instance.
+All the other members are there to provide information to those action functions, and the api of
+any other part is subject to arbritary breaking changes at any time.
+
+All action functions take the ctx and the go_toolchain as the only positional arguments, all
+other arguments even if mandator must be specified by name, to allow us to re-order and
+deprecate individual parameters over time.
+
+Inside the actions member are the following action emitting functions:
+
+* asm_
+* compile_
+* cover_
+* library_
+* link_
+* pack_
+
+asm
+~~~
+
+The asm function adds an action that runs ``go tool asm`` on a source file
+to produce an object.
+
+It does not return anything.
+
++--------------------------------+-----------------------------+-----------------------------------+
+| **Name**                       | **Type**                    | **Default value**                 |
++--------------------------------+-----------------------------+-----------------------------------+
+| :param:`ctx`                   | :type:`string`              | |mandatory|                       |
++--------------------------------+-----------------------------+-----------------------------------+
+| The current rule context, used to generate the actions.                                          |
++--------------------------------+-----------------------------+-----------------------------------+
+| :param:`go_toolchain`          | :type:`the go toolchain`    | |mandatory|                       |
++--------------------------------+-----------------------------+-----------------------------------+
+| This must be the same go toolchain object you got this function from.                            |
++--------------------------------+-----------------------------+-----------------------------------+
+| :param:`source`                | :type:`File`                | |mandatory|                       |
++--------------------------------+-----------------------------+-----------------------------------+
+| A source code artifact to assemble.                                                              |
++--------------------------------+-----------------------------+-----------------------------------+
+| :param:`hdrs`                  | :type:`list<File>`          | :value:`[]`                       |
```

Comment:
I think these types should be written as "list of Files" or "iterable of Files". In most (all?) cases, any iterable is accepted. Tuples and `depset`s should be allowed.